csrc/sycl/ops.cpp

```diff
@@ -1998,36 +2018,68 @@ template <typename T> void gemm_4bit_inference(int m, int n, int k, T * A,  unsi
     dpct::get_in_order_queue().submit(
       [&](sycl::handler &cgh) {
         /*
         DPCT1101:314: '8*16 + (16*(batch_size_warps-1))' expression was replaced with a value. Modify the code to use the original expression, provided in comments, if it is correct.
         */
-        sycl::local_accessor<T, 1> smem_A_acc_ct1(sycl::range<1>(176/*8*16 + (16*(batch_size_warps-1))*/), cgh);
+        
         /*
         DPCT1101:315: '2*batch_size_warps*16*32 + (2*16*(batch_size_warps-1))' expression was replaced with a value. Modify the code to use the original expression, provided in comments, if it is correct.
         */
+        
+        //__shared__ vars
+        sycl::local_accessor<T, 1> smem_A_acc_ct1(sycl::range<1>(176/*8*16 + (16*(batch_size_warps-1))*/), cgh);
         sycl::local_accessor<T, 1> smem_B_acc_ct1(sycl::range<1>(4192/*2*batch_size_warps*16*32 + (2*16*(batch_size_warps-1))*/), cgh);
         sycl::local_accessor<T, 1> smem_C_acc_ct1(sycl::range<1>(8*32), cgh);
 
         cgh.parallel_for(
           sycl::nd_range<3>(sycl::range<3>(1, 1, num_blocks) * sycl::range<3>(1, 1, 96), sycl::range<3>(1, 1, 96)), 
           [=](sycl::nd_item<3> item_ct1) {
-            kgemm_4bit_inference<T, 96>(m, n, k, A, B, absmax, out, lda, ldb, ldc, blocksize, item_ct1, smem_A_acc_ct1.get_pointer(), smem_B_acc_ct1.get_pointer(), smem_C_acc_ct1.get_pointer());
+            kgemm_4bit_inference<T, 96>(m, n, k, buff_A, buff_B, absmax, buff_out, lda, ldb, ldc, blocksize, item_ct1, smem_A_acc_ct1.get_pointer(), smem_B_acc_ct1.get_pointer(), smem_C_acc_ct1.get_pointer());
           });
       });
   }
   //kgemm_4bit_inference<T, 256><<< num_blocks, 256, 0, 0 >>>(m,  n,  k, A,  B, absmax, out, lda, ldb, ldc, blocksize);
   //kgemm_4bit_inference<T, 160><<< num_blocks, 160, 0, 0 >>>(m,  n,  k, A,  B, absmax, out, lda, ldb, ldc, blocksize);
   //kgemm_4bit_inference<T, 32><<< num_blocks, 32, 0, 0 >>>(m,  n,  k, A,  B, absmax, out, lda, ldb, ldc, blocksize);
+  //back memcpy
+  q_ct1.memcpy((T*)(A), (T*)(buff_A), size);
+  q_ct1.memcpy((unsigned char*)(B), (unsigned char*)(buff_B), size);
+  q_ct1.memcpy((T*)(out), (T*)(buff_out), size);
+  
 }
 
 template <typename T, int BITS> void gemm_4bit_inference_naive(int m, int n, int k, T * A,  unsigned char* B,  float *absmax, float *datatype, T * out,  int lda, int ldb, int ldc, int blocksize)
 {
 
 	int num_blocks = (m+3)/4;
+  dpct::device_ext &dev_ct1 = dpct::get_current_device();
+  sycl::queue &q_ct1 = dev_ct1.in_order_queue();
+  sycl::context ctx = q_ct1.get_context();
+    
+  int size = NUM_BLOCK;
+  *((T **)&buff_A) = sycl::malloc_device(size, A, ctx);
+  q_ct1.memcpy((T*)(buff_A), (T*)(A), size);
+  *(( unsigned char**)&buff_B) = sycl::malloc_device(size, B, ctx);
+  q_ct1.memcpy((unsigned char*)(buff_B), (unsigned char*)(B), size);
+  *((T **)&buff_out) = sycl::malloc_device(size, out, ctx);
+  q_ct1.memcpy((T*)(buff_out), (T*)(out), size);

```

Comment:
I'm just curious: is there a way to avoid all of this copying?

Reply:
Since each kernel call is initiated through ops , we have to dedicate some usm buffers for device processing. This is although a primary iteration , and i am thinking of ways to reduce the initiation duplication of device buffers.